docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,622 @@
+
+That’s a typical case where a service object comes in handy: you describe what the business logic is step by step (fetching models, checking permissions, and so on) and then you can take action depending on the outcome in a very simple way. All the error handling is done for you inside the service, and you can match its outcome in a very deterministic way. Input parameters are validated using a dedicated object called a contract and when a step fails, the service stops. That means that if a step runs, every previous step was successful.
+
+It doesn’t mean every situation should be handled by a service. A service encapsulates business logic for a given action, and is encouraged to rely on specialized objects. So it won’t replace models or libraries, for example. As said before, it really shines when used in a controller action, but it can be used anywhere. A background job is another good example.
+
+Another benefit immediately available with services, is that since all the logic related to a business action is encapsulated in it, you can call it from anywhere (console, SDK, controllers, specs, etc.) and you’ll always get the same outcome.
+
+It’s also quite easy to understand what’s happening in a service at a glance, since all the steps are listed sequentially. There’s also a matching system to handle the possible outcomes in a deterministic way, here again you can understand at a glance what a controller will do, for example.
+
+## Getting started
+
+Here’s a simplified service to update a user’s username which demonstrates most available steps:
+```ruby
+class User::UpdateUsername
+  include Service::Base
+
+  params do
+    attribute :id, :integer
+    attribute :username, :string
+
+    validates :id, presence: true
+    validates :username, presence: true, format: { with: /\A[a-zA-Z0-9]+\z/ }
+  end
+  model :user
+  policy :can_update_username
+  transaction do
+    step :update
+    step :log
+  end
+
+  private
+
+  def fetch_user(params:)
+    User.find_by(id: params[:id])
+  end
+
+  def can_update_username(guardian:, user:)
+    guardian.can_edit_username?(user)
+  end
+
+  def update(params:, user:)
+    user.update!(username: params[:username])
+  end
+
+  def log(guardian:, user:)
+    StaffActionLogger.new(guardian.user).log_username_change(
+      user,
+      user.username_before_last_save,
+      user.username,
+    )
+  end
+end
+```
+And here is an example on how you could call such a service:
+```ruby
+User::UpdateUsername.call(params: {id: 1, username: "username"}, guardian: Discourse.system_user.guardian)
+```
+
+Without knowing how services work, you can probably guess what’s happening here. Let’s dive in.
+
+## Steps
+
+### What’s a step?
```

Comment:
I think we need to explain how we handle external errors:

```ruby
def my_stem
  SomeLib.do_something # this might raise, how do I handle this?
end
```

Reply:
I think we have a part saying you shouldn’t raise as it will break the execution flow, but I guess I can expand a bit on this.
Right now, you would have to rescue the error yourself in your step, or make sure to call code that won’t raise. (raising can also be seen as that’s a bug, so we want to leave the exception pass through).

That said, I’ve been thinking about implementing another step (something Trailblazer has), which would allow rescuing exceptions. That way, it would be just a matter of doing something like this:
```ruby
rescue_from(MyException, :my_handler) do
  step :step_that_can_raise
  step :another_step_that_can_raise
end
```

Reply:
is handler a step? what does it have access to?

Reply:
That’s just an example. We could have a handler which could be a step, we could also have nothing in particular, and be able to match the exception in an outcome block, with something like this maybe:
```ruby
on_rescue_from(MyException) { |exception| do_something }
```

Reply:
I don't know if we need this complexity for now, I think we should just have a begin/rescue example for now.

We can always evolve to this pattern later, begin/rescue would be backwards compatible.

Reply:
TBH, the implementation should be pretty straightforward. But yes, for now, I can just expand on using a `rescue` block :+1: